File: dairy_app/dairy_app/report/milk_sale_summery_report/milk_sale_summery_report.py
# Copyright (c) 2024, Sigzen and contributors
# For license information, please see license.txt

import frappe

import frappe

def execute(filters=None):
    """
    Generates a milk sales summary report.
    """
    if not filters:
        filters = {}

    columns = [
        {"label": "Customer", "fieldname": "customer", "fieldtype": "data", "width": 150},
        {"label": "Posting Date", "fieldname": "posting_date", "fieldtype": "Date", "width": 150},
        {"label": "Liters Sold", "fieldname": "qty", "fieldtype": "Float", "width": 150},
        {"label": "Revenue", "fieldname": "amount", "fieldtype": "Currency", "width": 150},
    ]


    sales_invoices = frappe.get_all(
        "Sales Invoice",  
        fields=["customer", "posting_date"],
        filters=filters,
        order_by="posting_date asc"
    )

    data = []
    total_liters = 0
    total_revenue = 0

    for invoice in sales_invoices:
        child_items = frappe.get_all(
            "Sales Invoice Item",  
            fields=["qty", "amount"],
            filters={"parent": invoice.name}
        )
        
        liters = sum((item.qty or 0) for item in child_items)
        revenue = sum((item.amount or 0) for item in child_items)

        total_liters += liters
        total_revenue += revenue


        data.append({
            'customer':invoice.customer,
            "posting_date": invoice.posting_date,
            "qty": liters,  
            "amount": revenue,  
        })


    if total_liters > 0 or total_revenue > 0:
        data.append({
            "posting_date": "Total",
            "qty": total_liters,
            "amount": total_revenue,
        })

    return columns, data

File: dairy_app/dairy_app/report/milk_sale_summery_report/milk_sale_summery_report.js
// Copyright (c) 2024, Sigzen and contributors
// For license information, please see license.txt

frappe.query_reports["Milk Sale Summery Report"] = {
    "filters": [
        {
            "fieldname": "customer",
            "label": __("Customer"),
            "fieldtype": "Link",
            "options": "Customer",
            "reqd": 0
        },
        {
            "fieldname": "from_date",
            "label": __("From Date"),
            "fieldtype": "Date",
            "reqd": 0,
            "default": frappe.datetime.add_months(frappe.datetime.get_today(), -1)
        },
        {
            "fieldname": "to_date",
            "label": __("To Date"),
            "fieldtype": "Date",
            "reqd": 0,
            "default": frappe.datetime.get_today()
        }
    ]
};

